_source/_posts/2022-01-13-mobile-sso.md

```diff
@@ -0,0 +1,173 @@
+
+### Authentication prior to iOS 9
+
+Before the release of iOS 9 in 2015,  there were only two ways to authenticate through a web browser, and neither one was optimal. 
+
+* **`UIWebView`**. This approach shows web content in a `UIView`. The mobile app can intercept interactions with the `UIView`, hence it is not secure.
+
+* **Redirect to an external browser**. The mobile app can open a webpage in a separate browser app, and the browser app can redirect back to the mobile app after authentication. However,app switching is not a good experience for mobile users. In addition, the redirect back can fail if another app has registered the same URL scheme first.
+
+### Authentication in iOS 9 (2015)
+
+In 2015, Apple introduced [`SFSafariViewController`](https://developer.apple.com/documentation/safariservices/sfsafariviewcontroller), which displays an embedded Safari browser in the app. As an in-app browser tab, this interface lets the user stay in the app. Once the user interaction with the website is no longer needed, the browser tab can be closed and dismissed.  
+
+This solution was groundbreaking, because it is secure (the native app can neither peek into the embedded browser, nor alter its state) and it provides a better user experience by avoiding app switching. 
+
+`SFSafariViewController` shares cookies with the standalone Safari browser. This gives developers a way to implementSSO (Single Sign-On) with shared cookies..
+
+### iOS 11 changes (2017)
+
+Apple changed `SFSafariViewController` behavior to address privacy concerns with the release of iOS 11 in 2017. `SFSafariViewController` no longer shares cookies with the standalone Safari browser. Thus, a website can no longer determine that the user on `SFSafariViewController` is the same user on Safari, even though the view controller and the browser are open on the same device. 
+
+Apple understood that  the `SFSafariViewController` behavior change would break SSO, since single sign-on relies on the ability to share cookies. Instead, Apple had to introduce `SFAuthenticationSession` as a workaround, where `SFAuthenticationSession` would share a persistent cookie with Safari. 
+
+### Iterative changes in iOS 12 (2018)
+
+In 2018, Apple deprecated `SFAuthenticationSession`, but introduced [`ASWebAuthenticationSession`](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession) as a revised solution. 
+
+Both `SFAuthenticationSession` and `ASWebAuthenticationSession` are designed specifically for OAuth 2.0 authentication, not for showing general web content. To avoid abuse, when `SFAuthenticationSession` or `ASWebAuthenticationSession` are used, Apple always displays the prompt we saw earlier., The prompt is designed explicitly for user sign-in, and indicates that cookies are shared. 
```

Comment:
```suggestion
Both `SFAuthenticationSession` and `ASWebAuthenticationSession` are designed specifically for OAuth 2.0 authentication, not for showing general web content. To avoid abuse, when `SFAuthenticationSession` or `ASWebAuthenticationSession` are used, Apple always displays the prompt we saw earlier. The prompt is designed explicitly for user sign-in, and indicates that cookies are shared. 
```
removed typo - comma